aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -554,15 +816,165 @@ module aptos_framework::delegation_pool {
+    fun get_delegator_active_shares(pool: &DelegationPool, delegator: address): u128 {
+        pool_u64::shares(&pool.active_shares, delegator)
+    }
+
+    /// Get the pending inactive share amount of the delegator.
+    fun get_delegator_pending_inactive_shares(pool: &DelegationPool, delegator: address): u128 {
+        pool_u64::shares(pending_inactive_shares_pool(pool), delegator)
+    }
+
+    /// Get the used voting power of a voter on a proposal.
+    fun get_used_voting_power(governance_records: &GovernanceRecords, voter: address, proposal_id: u64): u64 {
+        let votes = &governance_records.votes;
+        let key = VotingRecordKey {
+            voter,
+            proposal_id,
+        };
+        *smart_table::borrow_with_default(votes, key, &0)
+    }
+
+    /// Borrow the mutable used voting power of a voter on a proposal.
+    inline fun borrow_mut_used_voting_power(governance_records: &mut GovernanceRecords, voter: address, proposal_id: u64): &mut u64 {
+        let votes = &mut governance_records.votes;
+        let key = VotingRecordKey {
+            proposal_id,
+            voter,
+        };
+        smart_table::borrow_mut_with_default(votes, key, 0)
+    }
+
+    /// Get the latest VoteDelegation of a delegator.
+    /// The 2nd bool return value is true when VoteDelegation has already been created before calling this function.
+    /// The 3rd bool return value is true when the latest VoteDelegation is different from the existing VoteDelegation.
+    fun get_latest_delegator_vote_delegation(pool: &DelegationPool, governance_records: &GovernanceRecords, delegator: address): (VoteDelegation, bool, bool) {
+        let pool_address = get_pool_address(pool);
+        let locked_until_secs = stake::get_lockup_secs(pool_address);
+        let vote_delegation = &governance_records.vote_delegation;
+        // By default, a delegator's delegated voter is itself.
+        if (!smart_table::contains(vote_delegation, delegator)) {
+            return (VoteDelegation {
+                voter: delegator,
+                last_locked_until_secs: locked_until_secs,
+                pending_voter: delegator,
+            }, false, false)
+        };
+        let vote_delegation = *smart_table::borrow(vote_delegation, delegator);
+        let needs_update = false;
+        if (vote_delegation.last_locked_until_secs < locked_until_secs &&
+            vote_delegation.voter != vote_delegation.pending_voter) {
+            vote_delegation.voter = vote_delegation.pending_voter;
+            needs_update = true;
+        };
+        (vote_delegation, true, needs_update)
+    }
+
+    /// Update VoteDelegation of a delegator to up-to-date then borrow_mut it.
+    fun update_and_borrow_mut_delegator_vote_delegation(
+        pool: &DelegationPool,
+        governance_records :&mut GovernanceRecords,
+        delegator: address
+    ): &mut VoteDelegation {
+        let (latest_vote_delegation, is_existed, needs_update) =
```

Comment:
This looks funky to me because:
1. get_latest_delegator_vote_delegation is only ever called here in this function
2. get_latest_delegator_vote_delegation borrows multiple, makes a copy of the VotingDelegation struct, and returns it here. The struct is then modified further here

I think you can just merge the logic in get_latest_delegator_vote_delegation here.

Reply:
Make sense. `get_latest_delegator_vote_delegation` was used in a view function so I want to keep it read-only. But the view function has to sync first. Will merge the logic.